### PR TITLE
ENH: Performance improvements for `linear_sum_assignment`

### PR DESCRIPTION
**Pull request by pmla**
#### What does this implement/fix?

Some small improvements to the `linear_sum_assignment` method:

- The Jonker-Volgenant algorithm does not require a non-negative cost matrix.
- The `remaining` vector can be re-used between function calls.
- The contents of `_lsap.py` has been moved into `_lsap.c`. This removes the overhead of a function call.
- The cost matrix is only copied if it needs modifying.

#### Additional information

The benchmarking output on my computer (compared to main):
```
[ 50.00%] · For scipy commit 0051e39a (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.9-Cython0.29.21-numpy-pybind11-pytest-pythran
[ 75.00%] ··· optimize_lap.LinearAssignment.time_evaluation                                                                 ok
[ 75.00%] ··· ============ ============= ============ ============= ============= =============
              --                                        cost_type                              
              ------------ --------------------------------------------------------------------
                 shape        uniform      spatial     logarithmic     integer        binary   
              ============ ============= ============ ============= ============= =============
               (100, 100)     312±40μs     395±60μs      366±30μs      321±20μs      160±5μs   
               (200, 200)   1.29±0.06ms   1.99±0.2ms   1.42±0.09ms   1.24±0.07ms     532±5μs   
               (300, 300)   3.30±0.09ms    4.74±1ms     3.36±0.2ms    3.31±0.2ms   1.26±0.07ms 
               (400, 400)    6.20±0.4ms   10.6±0.6ms    6.40±0.2ms    5.93±0.4ms    2.16±0.1ms 
               (100, 200)     223±10μs     224±20μs      228±10μs      220±10μs      280±20μs  
               (200, 400)     781±40μs     783±40μs      781±40μs      794±40μs    1.04±0.05ms 
               (300, 600)   1.61±0.09ms   1.79±0.2ms   1.56±0.03ms   1.63±0.04ms   2.15±0.04ms 
               (400, 800)   2.73±0.07ms   2.90±0.1ms   2.75±0.05ms   2.93±0.09ms   3.80±0.06ms 
               (200, 100)     229±10μs     234±20μs      223±6μs       243±6μs       292±3μs   
               (400, 200)     851±30μs     877±30μs      833±30μs     1.13±0.1ms    1.39±0.1ms 
               (600, 300)   1.82±0.03ms   2.07±0.2ms   1.86±0.09ms    2.53±0.3ms    3.20±0.3ms 
               (800, 400)    3.85±0.1ms   3.95±0.1ms    4.31±0.8ms    5.09±0.5ms    6.17±0.6ms 
              ============ ============= ============ ============= ============= =============

[ 75.00%] · For scipy commit 54f23d4c (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.9-Cython0.29.21-numpy-pybind11-pytest-pythran..
[ 75.00%] ·· Benchmarking virtualenv-py3.9-Cython0.29.21-numpy-pybind11-pytest-pythran
[100.00%] ··· optimize_lap.LinearAssignment.time_evaluation                                                                 ok
[100.00%] ··· ============ ============= ============= ============= ============= ============
              --                                        cost_type                              
              ------------ --------------------------------------------------------------------
                 shape        uniform       spatial     logarithmic     integer       binary   
              ============ ============= ============= ============= ============= ============
               (100, 100)     392±40μs      441±90μs      388±30μs      341±20μs     185±1μs   
               (200, 200)   1.47±0.04ms    2.24±0.2ms   1.60±0.09ms    1.46±0.1ms    633±3μs   
               (300, 300)   3.82±0.05ms     5.38±1ms     3.91±0.3ms   3.93±0.06ms   1.76±0.2ms 
               (400, 400)    6.85±0.5ms    11.1±0.4ms    6.71±0.3ms    6.86±0.1ms   3.08±0.3ms 
               (100, 200)     265±10μs      262±20μs      265±7μs       264±7μs      311±3μs   
               (200, 400)     915±30μs      893±50μs      922±40μs     1.26±0.2ms   1.47±0.2ms 
               (300, 600)   1.95±0.06ms    2.28±0.2ms   1.97±0.05ms    2.82±0.4ms   3.42±0.4ms 
               (400, 800)   3.55±0.07ms    3.73±0.1ms   3.51±0.05ms    5.27±0.6ms   6.21±0.5ms 
               (200, 100)     276±10μs      277±20μs      273±8μs       279±8μs      327±5μs   
               (400, 200)   1.01±0.03ms   1.03±0.03ms     991±30μs     1.31±0.1ms   1.54±0.2ms 
               (600, 300)   2.14±0.04ms   2.20±0.04ms   2.11±0.02ms    2.87±0.3ms   3.47±0.3ms 
               (800, 400)   4.33±0.08ms    4.47±0.1ms    4.37±0.1ms    5.96±0.7ms   6.98±0.6ms 
              ============ ============= ============= ============= ============= ============

       before           after         ratio
     [54f23d4c]       [0051e39a]
-         311±3μs         280±20μs     0.90  optimize_lap.LinearAssignment.time_evaluation((100, 200), 'binary')
-         327±5μs          292±3μs     0.89  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'binary')
-     4.33±0.08ms       3.85±0.1ms     0.89  optimize_lap.LinearAssignment.time_evaluation((800, 400), 'uniform')
-         279±8μs          243±6μs     0.87  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'integer')
-     3.82±0.05ms      3.30±0.09ms     0.86  optimize_lap.LinearAssignment.time_evaluation((300, 300), 'uniform')
-     2.14±0.04ms      1.82±0.03ms     0.85  optimize_lap.LinearAssignment.time_evaluation((600, 300), 'uniform')
-     3.93±0.06ms       3.31±0.2ms     0.84  optimize_lap.LinearAssignment.time_evaluation((300, 300), 'integer')
-     1.01±0.03ms         851±30μs     0.84  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'uniform')
-         633±3μs          532±5μs     0.84  optimize_lap.LinearAssignment.time_evaluation((200, 200), 'binary')
-        991±30μs         833±30μs     0.84  optimize_lap.LinearAssignment.time_evaluation((400, 200), 'logarithmic')
-     1.95±0.06ms      1.61±0.09ms     0.83  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'uniform')
-         273±8μs          223±6μs     0.82  optimize_lap.LinearAssignment.time_evaluation((200, 100), 'logarithmic')
-     1.97±0.05ms      1.56±0.03ms     0.79  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'logarithmic')
-     3.51±0.05ms      2.75±0.05ms     0.78  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'logarithmic')
-      3.73±0.1ms       2.90±0.1ms     0.78  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'spatial')
-      1.76±0.2ms      1.26±0.07ms     0.71  optimize_lap.LinearAssignment.time_evaluation((300, 300), 'binary')
-      1.47±0.2ms      1.04±0.05ms     0.71  optimize_lap.LinearAssignment.time_evaluation((200, 400), 'binary')
-      3.08±0.3ms       2.16±0.1ms     0.70  optimize_lap.LinearAssignment.time_evaluation((400, 400), 'binary')
-      3.42±0.4ms      2.15±0.04ms     0.63  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'binary')
-      6.21±0.5ms      3.80±0.06ms     0.61  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'binary')
-      2.82±0.4ms      1.63±0.04ms     0.58  optimize_lap.LinearAssignment.time_evaluation((300, 600), 'integer')
-      5.27±0.6ms      2.93±0.09ms     0.56  optimize_lap.LinearAssignment.time_evaluation((400, 800), 'integer')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

@peterbell10

### CODE REVIEW

**Comment by tylerjereddy** [comment]
@pmla do you have time to fix the merge conflicts here?

**Comment by pmla** [comment]
@tylerjereddy I have made a mess of my first attempt at fixing the conflicts. I will fix them later today.

**Comment by pmla** [comment]
@tylerjereddy conflicts are fixed. Errors are related to the problem described in #16273.

**Review by tylerjereddy — APPROVED** [review]
Looks like the code is pretty well tested and the compiled code patch diff is 100 %. Since there are no testing changes whatsoever, and a performance improvement is reported with a net decrease in LOC, I don't think I have a reason not to merge.  The diff seems pretty clean to me. An algorithm expert reviewing would be good, but seems safe enough.

**Comment by tylerjereddy** [comment]
thanks

**Comment by pmla** [comment]
Thanks Tyler.

File: scipy/optimize/_lsap.c [hunk 1]
```diff
@@ -96,26 +101,96 @@ calculate_assignment(PyObject* self, PyObject* args)
...
     NULL,
     NULL,
     NULL,
     NULL,
 };
 
-PyMODINIT_FUNC
-PyInit__lsap_module(void)
+PyObject*
+PyInit__lsap(void)
```
Comment (eli-schwartz) [inline scipy/optimize/_lsap.c:193]: A bit of a subtle issue here, but this merge conflicts resolution broke using `-fvisibility=hidden` which is a neat optimization and guarantees correctness as described here: https://docs.python.org/3/extending/extending.html#the-module-s-method-table-and-initialization-function

It's an easy fix, and #16541 does it.

**Pull request merged**